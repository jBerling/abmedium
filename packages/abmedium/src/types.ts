import {
  simName,
  symName,
  seqName,
  disName,
  strName,
  nilName,
  numName,
  refName,
} from "./constants";

export type Nil = null;
export type Str = string;
export type Num = number;
export type Sim = [typeof simName, NodeValue[]];
export type Dis = [
  typeof disName,
  {
    expected: NodeValue | undefined;
    actual: NodeValue | undefined;
    to: NodeValue;
  }
];
export type Sym = [typeof symName, string];
export type Ref = [typeof refName, Label];
export type Seq = [typeof seqName, Array<Label>];
export type Scalar = Nil | Str | Num | Sym | Ref;
export type CompoundValue = Sim | Dis | Seq;
export type NodeValue = Scalar | CompoundValue;

export type NodeValueType =
  | typeof symName
  | typeof simName
  | typeof seqName
  | typeof strName
  | typeof nilName
  | typeof numName
  | typeof disName
  | typeof refName;

export type Label = string | number;

type LayerRecord<T> = Record<Label, T | SubLayerRecord<T>>;
export interface SubLayerRecord<T> extends LayerRecord<T> {}
export type Layer = LayerRecord<NodeValue>;

export type Metalayer = Record<Label, NodeValue>;

export type LayerWithSublayers = [Label, ViewStack];

export type ViewStack = (Label | LayerWithSublayers)[];

export type Projection = {
  nodes: Record<Label, NodeValue>;
  metadata: Record<Label, Metalayer>;
  simultaneities: Record<Label, Sim>;
  disagreements: Record<Label, Dis>;
};

export type PresentationNode<R = Scalar> = {
  value: NodeValue;
  items?: R[];
  label: Label;
  parent?: Label;
  pos?: number;
  dis?: NodeValue;
  sim?: NodeValue;
  metadata: Record<string, NodeValue>;
  disagreement?: {
    expected?: NodeValue;
    actual?: NodeValue;
    to: NodeValue;
  };
  simultaneities?: NodeValue[];
};

export type NodePresenter<R = Scalar> = (node: PresentationNode<R>) => R;
