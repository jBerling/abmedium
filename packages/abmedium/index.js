const {
  layer,
  isLayer,
  document,
  isDocument,
  disagreement,
  sym,
  sim,
  str,
  num,
  seq,
  seqItems,
  nil,
  mapping,
  valtype,
  lengthOf,
  editvalOf,
  equal,
} = require('./src/core');

const proj = require('./src/proj');
const treeOf = require('./src/tree-of');
const nodes = require('./src/nodes');
const layers = require('./src/layers');

module.exports = {
  layer,
  isLayer,
  document,
  isDocument,
  disagreement,
  sym,
  sim,
  str,
  num,
  seq,
  seqItems,
  nil,
  mapping,
  valtype,
  lengthOf,
  editvalOf,
  equal,

  proj,
  treeOf,
  nodes,
  layers,
};
