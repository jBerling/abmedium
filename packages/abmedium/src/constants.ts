export const symName = "sym";
export const numName = "num";
export const seqName = "seq";
export const strName = "str";
export const nilName = "nil";
export const refName = "ref";

export const scalarTypeNames: [
  typeof nilName,
  typeof strName,
  typeof numName,
  typeof symName,
  typeof refName
] = [nilName, strName, numName, symName, refName];
