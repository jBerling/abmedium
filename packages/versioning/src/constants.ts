export const mainDir = ".abv";
export const objectsDir = ".objects";
export const timestampsLayer = "timestamps";
export const counter = "counter";
export const head = "head";
export const viewStack = ".viewstack";
