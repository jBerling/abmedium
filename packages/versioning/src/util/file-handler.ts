import {
  readFile as pipeReadFile,
  writeFile as pipeWriteFile,
  unlink as pipeUnlink,
  readDir as pipeReadDir,
  ensureDir as pipeEnsureDir,
  glob as pipeGlob,
} from "./pipable-operators";

import {
  readFile,
  writeFile,
  unlink,
  readDir,
  ensureDir,
  glob,
} from "./creation-operators";
import { resolve } from "path";
import { FileHandler } from "./types";

export const fileHandler: FileHandler = {
  resolve,

  readFile,
  writeFile,
  unlink,
  readDir,
  ensureDir,
  glob,

  pipable: {
    readFile: pipeReadFile,
    writeFile: pipeWriteFile,
    unlink: pipeUnlink,
    readDir: pipeReadDir,
    ensureDir: pipeEnsureDir,
    glob: pipeGlob,
  },
};
