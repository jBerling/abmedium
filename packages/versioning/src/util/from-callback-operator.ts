import { Observable, Observer } from "rxjs";

type Next<Out> = (out: Out) => void;

type Callback<In, Out> = (
  sourceValue: In,
  next: Next<Out>,
  errorHandler: (err: Error | string) => void
) => void;

export default <T, R>(cb: Callback<T, R>) => (source: Observable<T>) =>
  new Observable((observer: Observer<R>) => {
    let items = 0;
    let completed = 0;
    let complete = () => {};

    const subscription = source.subscribe({
      next(x: T) {
        items += 1;
        cb(
          x,
          (out) => {
            observer.next(out);
            completed += 1;
            complete();
          },
          (error) => {
            observer.error(error);
          }
        );
      },
      error(err) {
        observer.error(err);
      },
      complete() {
        complete = () => {
          if (items === completed) observer.complete();
        };
        complete();
      },
    });

    return function teardown() {
      subscription.unsubscribe();
    };
  });
