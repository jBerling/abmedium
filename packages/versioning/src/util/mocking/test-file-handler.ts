import { Observable } from "rxjs";
import { FileHandler } from "../types";
import fromCallback from "../../util/from-callback-operator";
import { Minimatch } from "minimatch";

type Directory<T> = Record<string, string | T>;
export type TestFiles = Directory<Directory<Directory<string>>>;

const pathSegments = (path: string): string[] =>
  path.split("/").filter((p) => p);

const get = (directory: any, [head, ...rest]: string[]) => {
  if (!head) throw new Error("head was undefined");
  if (!rest || !rest.length) return directory[head];
  return get(directory[head], rest);
};

const warnAboutFilePath = (path: string) => {
  if (path.startsWith("/")) {
    console.warn("Actual file path used?", path);
  }
};

const getFile = (files: TestFiles, path: string) => {
  warnAboutFilePath(path);
  const res = get(files, pathSegments(path));
  if (!res || typeof res !== "string") {
    throw new Error("No file named " + path);
  }
  return { filename: path, data: res };
};

const getDir = (files: TestFiles, path: string) => {
  warnAboutFilePath(path);

  const res = path ? get(files, pathSegments(path)) : files;

  if (!res || typeof res !== "object") {
    throw new Error("No directory named " + path);
  }
  return res;
};

const createDir = (files: TestFiles, path: string) => {
  warnAboutFilePath(path);

  const segments = pathSegments(path);
  const parent = getDir(
    files,
    segments.slice(0, segments.length - 1).join("/")
  );

  parent[segments[segments.length - 1]] = {};
};

const createFile = (files, path: string, data: string): void => {
  warnAboutFilePath(path);

  const segments = pathSegments(path);
  const parent = getDir(
    files,
    segments.slice(0, segments.length - 1).join("/")
  );

  parent[segments[segments.length - 1]] = data;
};

const exists = (files: TestFiles, path: string): boolean => {
  warnAboutFilePath(path);

  try {
    const res = get(
      files,
      path.split("/").filter((p) => p)
    );
    return !!res;
  } catch (err) {
    return false;
  }
};

const filenames = (files: TestFiles): string[] => {
  const names: string[] = [];

  const walk = (dir: TestFiles, parentPath: string): void => {
    for (const key of Object.keys(dir)) {
      const data = dir[key];
      const path = parentPath ? parentPath + "/" + key : key;
      if (typeof data === "string") {
        names.push(path);
      } else if (typeof data === "object") {
        walk(data, path);
      }
    }
  };

  walk(files, "");

  return names;
};

export const testFileHandler = (files: TestFiles): FileHandler => ({
  resolve: (...pathSegments: string[]): string =>
    pathSegments.filter((s) => s).join("/"),

  readFile: (fileName) =>
    new Observable((subscriber) => {
      try {
        subscriber.next(getFile(files, fileName).data);
        subscriber.complete();
      } catch (err) {
        subscriber.error(err);
      }
    }),
  writeFile: (fileName, data) =>
    new Observable((subscriber) => {
      try {
        createFile(files, fileName, data);
        subscriber.next();
        subscriber.complete();
      } catch (err) {
        subscriber.error(err);
      }
    }),
  unlink: (fileName: string) =>
    new Observable((subscriber) => {
      try {
        const segments = pathSegments(fileName);
        const dirPath = segments.slice(0, segments.length - 1).join("/");
        const dir = getDir(files, dirPath);
        if (dir) dir[segments[segments.length - 1]] = undefined;
        subscriber.next();
        subscriber.complete();
      } catch (err) {
        subscriber.error(err);
      }
    }),
  readDir: (dirName) =>
    new Observable((subscriber) => {
      try {
        subscriber.next(Object.keys(getDir(files, dirName)));
        subscriber.complete();
      } catch (err) {
        subscriber.error(err);
      }
    }),
  ensureDir: (dirName) =>
    new Observable((subscriber) => {
      try {
        if (!exists(files, dirName)) createDir(files, dirName);
        subscriber.next();
        subscriber.complete();
      } catch (err) {
        subscriber.error(err);
      }
    }),
  glob: (pattern, options = {}) =>
    new Observable((subscriber) => {
      try {
        const mm = new Minimatch(pattern, options);
        subscriber.next(filenames(files).filter((fname) => mm.match(fname)));
        subscriber.complete();
      } catch (err) {
        subscriber.error(err);
      }
    }),

  pipable: {
    readFile: () =>
      fromCallback((filename, next, errorHandler) => {
        try {
          next(getFile(files, filename).data);
        } catch (err) {
          console.error(err);
          errorHandler(err);
        }
      }),

    writeFile: () =>
      fromCallback(([fileName, data], next, errorHandler) => {
        try {
          createFile(files, fileName, data);
          next();
        } catch (err) {
          console.error(err);
          errorHandler(err);
        }
      }),

    unlink: () =>
      fromCallback((fileName, next, errorHandler) => {
        try {
          const segments = pathSegments(fileName);
          const dirPath = segments.slice(0, segments.length - 1).join("/");
          const dir = getDir(files, dirPath);
          if (dir) dir[segments[segments.length - 1]] = undefined;
          next();
        } catch (err) {
          console.error(err);
          errorHandler(err);
        }
      }),

    readDir: () =>
      fromCallback((dirname, next, errorHandler) => {
        try {
          next(Object.keys(getDir(files, dirname)));
        } catch (err) {
          console.error(err);
          errorHandler(err);
        }
      }),

    ensureDir: () =>
      fromCallback((dirName, next, errorHandler) => {
        try {
          if (!exists(files, dirName)) createDir(files, dirName);
          next();
        } catch (err) {
          console.error(err);
          errorHandler(err);
        }
      }),

    glob: (options = {}) =>
      fromCallback((pattern: string, next, errorHandler) => {
        const mm = new Minimatch(pattern, options);
        try {
          next(filenames(files).filter((fname) => mm.match(fname)));
        } catch (err) {
          errorHandler(err);
        }
      }),
  },
});
