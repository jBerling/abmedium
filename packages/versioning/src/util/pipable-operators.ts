import fs from "fs";
import glob_ from "glob";
import fromCallback from "./from-callback-operator";

export const readFile = (encoding: "utf8") =>
  fromCallback<string, string>((fileName, next, errorHandler) => {
    fs.readFile(fileName, encoding, (err, data) => {
      if (err) errorHandler(err);
      else next(data);
    });
  });

export const writeFile = (encoding: "utf8") =>
  fromCallback<[string, string], void>(
    ([fileName, data], next, errorHandler) => {
      fs.writeFile(fileName, data, encoding, (err) => {
        if (err) errorHandler(err);
        else next();
      });
    }
  );

export const unlink = () =>
  fromCallback<string, void>((filename, next) => {
    fs.unlink(filename, () => next());
  });

export const readDir = () =>
  fromCallback<string, string[]>((filename, next, errorHandler) => {
    fs.readdir(filename, (err, files) => {
      if (err) errorHandler(err);
      else next(files);
    });
  });

export const ensureDir = () =>
  fromCallback<string, void>((dirname, next, errorHandler) => {
    try {
      if (fs.existsSync(dirname)) {
        next();
      } else {
        fs.mkdir(dirname, (err) => {
          if (err) errorHandler(err);
          else next();
        });
      }
    } catch (err) {
      errorHandler(err);
    }
  });

export const glob = () =>
  fromCallback<string, string[]>((filePattern, next, errorHandler) => {
    try {
      glob_(filePattern, {}, (err, matches) => {
        if (err) errorHandler(err);
        else next(matches);
      });
    } catch (err) {
      errorHandler(err);
    }
  });
