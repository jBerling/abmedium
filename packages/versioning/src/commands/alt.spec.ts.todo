import { TestScheduler } from "rxjs/testing";
import { FileHandler } from "../util/types";
import { testFileHandler, TestFiles } from "../util/mocking/test-file-handler";
import alt, { layerStacking, layerByPath } from "./alt";
import { document, num, seq, nil, layer } from "@abrovink/abmedium";
import { mainDir, objectsDir, counter, viewStack, head } from "../constants";

const testScheduler = () =>
  new TestScheduler((actual, expected) => {
    expect(JSON.parse(JSON.stringify(actual))).toEqual(
      JSON.parse(JSON.stringify(expected))
    );
  });

const run = (fn) => () => testScheduler().run(fn);

const hasher = ({ fileName }: { fileName: string; data: string }): string =>
  `<${fileName.split(".")[0]}>.${fileName.split(".")[1]}`;

const now = () => "2020-07-20T18:36Z";

describe("abv alt", () => {
  let fileHandler: FileHandler;
  let archive: TestFiles;

  beforeEach(() => {
    archive = {
      [mainDir]: {
        [viewStack]: `["sub-layer", "alt-layer"]`,
        [objectsDir]: {},
        foo: JSON.stringify(
          document({
            [counter]: num(2),
            0: seq(nil, 1),
            1: "ver1.md",
            [head]: num(0),
          }),
          null,
          4
        ),
      },
      "foo.md": "# Foo\n",
    };
    fileHandler = testFileHandler(archive);
  });

  test(
    "adds alt version",
    run(({ expectObservable, flush }) => {
      expectObservable(
        alt({
          fileHandler,
          hasher,
          now,
          archiveName: "",
          id: "foo",
        })
      ).toBe("(a|)", { a: undefined });

      flush();

      expect(archive).toMatchSnapshot();
    })
  );
});

// TODO move
describe("layerStacking", () => {
  test("[]", () => expect(layerStacking([])).toEqual([]));
  test(`["a", "b", "c"]`, () => {
    expect(layerStacking(["a", "b", "c"])).toEqual([["a"], ["b"], ["c"]]);
  });
  test(`["a", ["b", ["b.1", "b.2"]], ["c", [["c.1", ["c.1.1", "c.1.2"]], "c.2"]]]`, () => {
    expect(
      layerStacking([
        "a",
        ["b", ["b.1", "b.2"]],
        ["c", [["c.1", ["c.1.1", "c.1.2"]], "c.2"]],
      ])
    ).toEqual([
      ["a"],
      ["b"],
      ["b", "b.1"],
      ["b", "b.2"],
      ["c"],
      ["c", "c.1"],
      ["c", "c.1", "c.1.1"],
      ["c", "c.1", "c.1.2"],
      ["c", "c.2"],
    ]);
  });
});

// TODO move
describe("layerByPath", () => {
  test("...", () => {
    expect(
      layerByPath(
        layer({
          a: layer({ b: layer({ c: layer({ name: "right" }) }) }),
          c: layer({ name: "wrong" }),
        }),
        ["a", "b", "c"]
      )
    ).toEqual(layer({ name: "right" }));
  });
});
