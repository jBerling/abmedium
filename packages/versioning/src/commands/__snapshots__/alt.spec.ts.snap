// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`abv alt adds alt version 1`] = `
Object {
  ".abv": Object {
    ".objects": Object {
      "<foo>.md": "# Foo
",
    },
    ".viewstack": "[\\"alt-layer\\"]",
    "foo": "{
    \\"0\\": [
        \\"seq\\",
        [
            null,
            1
        ]
    ],
    \\"1\\": \\"ver1.md\\",
    \\"counter\\": 4,
    \\"head\\": 0,
    \\"__document__\\": true,
    \\"__layer__\\": true,
    \\"alt-layer\\": {
        \\"2\\": [
            \\"mapping\\",
            {
                \\"to\\": [
                    \\"seq\\",
                    [
                        0,
                        3
                    ]
                ]
            }
        ],
        \\"3\\": [
            \\"mapping\\",
            {
                \\"to\\": \\"<foo>.md\\"
            }
        ],
        \\"__layer__\\": true,
        \\"head\\": [
            \\"mapping\\",
            {
                \\"from\\": 0,
                \\"to\\": 2
            }
        ],
        \\"timestamps\\": {
            \\"2\\": [
                \\"mapping\\",
                {
                    \\"to\\": \\"2020-07-20T18:36Z\\"
                }
            ],
            \\"__layer__\\": true
        }
    }
}",
  },
  "foo.md": "# Foo
",
}
`;
